### ISSUE TITLE
Incomplete tests: fault injection

### ISSUE THREAD

**Issue by kyessenov** [istio/old_pilot_repo#1419]
Context: [production readiness proposal and plan](https://goo.gl/pmjuYg)

The feature “_fault injection_” is identified as incomplete test coverage. As a measure to reach Istio production quality in 0.3, this issue is opened to track the progress of providing sufficient test coverage for the feature in question.

This issue is initially assigned to the working group leads.

**Recommendations**:
- Come up with a test plan for how the feature should be tested. Identify and document (in this issue, or a separate document this issue will link to) test cases and whether each will be covered by an integration test (with some level of mocking and/or starting up selected components in test environment), or end to end test (aka system test, which brings up the whole Istio to exercise the feature; currently live in the istio/istio repo).
- If you expect the test plan is complex and controversial, consider having it formally reviewed.
- State an ETA in the issue, and assign this to a developer who will execute the test plan.
- Execute the test plan as the highest priority for 0.3.
- For any bug or defect identified by the new test cases, file a new issue and link it from this one.
- Close this issue only when the test plan is completely executed, and all additional issues identified in step 3 are fixed.

If you see a risk of not fixing this issue by end of 0.3, please contact istio-production-quality@googlegroups.com as soon as possible. We will also periodically ping the issue for update.

**Comment by rshriram** [comment]
What tests are needed? We have e2e tests in both pilot and istio

**Comment by kyessenov** [comment]
Does fault injection work with ingress?

**Comment by rshriram** [comment]
Thats a separate *feature*, tracked under ingress bug.

**Comment by cmluciano** [comment]
I'm fine to work on this @andraxylia , but I thought that we were blocked on Ingress issues until the new API stuff comes out

**Comment by andraxylia** [comment]
@rshriram Fault injection is configured via routing rules, so it should not be blocked on Ingress work, but in any case, if this is not currently working, we can have the tests ready and enable them later. Do you plan to change them as not to be configured via routing rules anymore, as part of the better ingress?

Currently, we have coverage for Fault Injection with Timeouts, in UT, pilot integration and e2e tests. What is missing is:
- Fault injection with Aborts for http, grpc, tcp - missing from UT, integration, e2e.
- Fault injection with Timeouts and Aborts for Ingress - missing from UT, integration, e2e.

At a minimum, we can add coverage in integration.

**Comment by rshriram** [comment]
On Tue, Oct 31, 2017 at 11:43 AM Andra Cismaru <notifications@github.com>
wrote:

> @rshriram <https://github.com/rshriram> Fault injection is configured via
> routing rules, so it should not be blocked on Ingress work, but in any
> case, if this is not currently working, we can have the tests ready and
> enable them later. Do you plan to change them as not to be configured via
> routing rules anymore, as part of the better ingress?
>
> Currently, we have coverage for Fault Injection with Timeouts, in UT,
> pilot integration and e2e tests. What is missing is:
>
>    - Fault injection with Aborts for http, grpc, tcp - missing from UT,
>    integration, e2e.
>
>
We just need this. Nothing wise.


>    -
>    - Fault injection with Timeouts and Aborts for Ingress - missing from
>    UT, integration, e2e.
>
>
we know that fault injection does not work for ingress. There is a
structural issue here (hard to explain. Will be apparent when you read the
ingress.go code).

**Comment by kyessenov** [comment]
Issue moved to [istio/istio #1474](https://github.com/istio/istio/issues/1474) via [**ZenHub**](https://www.zenhub.com/)
